src/index.ts

```diff
@@ -1,33 +1,63 @@
-export interface Env {
-  // Example binding to KV. Learn more at https://developers.cloudflare.com/workers/runtime-apis/kv/
-  // MY_KV_NAMESPACE: KVNamespace;
-  //
-  // Example binding to Durable Object. Learn more at https://developers.cloudflare.com/workers/runtime-apis/durable-objects/
-  // MY_DURABLE_OBJECT: DurableObjectNamespace;
-  //
-  // Example binding to R2. Learn more at https://developers.cloudflare.com/workers/runtime-apis/r2/
-  // MY_BUCKET: R2Bucket;
-  //
-  // Example binding to a Service. Learn more at https://developers.cloudflare.com/workers/runtime-apis/service-bindings/
-  // MY_SERVICE: Fetcher;
-  //
-  // Example binding to a Queue. Learn more at https://developers.cloudflare.com/queues/javascript-apis/
-  // MY_QUEUE: Queue;
-}
-
-export default {
-  async fetch (request: Request, env: Env, ctx: ExecutionContext): Promise<Response> {
-    return new Response('Hello World!')
-  }
-}
+import { Env } from "./env";
+import { Hono } from "hono";
+import { Error, ServiceAuthType } from "./types";
+import Services from "./services";
+import { streamResponse } from "./utils";
+
+const app = new Hono<{ Bindings: Env }>();
+app
+  .all("*",
+    async (c) => {
+      const xGatewayService = c.req.raw.headers.get("x-gateway-service");
+      const xGatewayServiceHost = c.req.raw.headers.get("x-gateway-service-host");
+      const xGatewayServiceAuthType = c.req.raw.headers.get("x-gateway-service-auth-type");
+      const xGatewayServiceAuthKey = c.req.raw.headers.get("x-gateway-service-auth-key");
+      const xGatewayServiceToken = c.req.raw.headers.get("x-gateway-service-token");
+
+      if (!xGatewayService) {
+        return c.json<Error>({
+          error: "x-gateway-service header not found!"
+        }, { status: 400 });
+      }
+
+      const service = new Services(c).get(xGatewayService);
+
+      if (!service) {
+        return c.json<Error>({
+          error: "service not found!"
+        }, { status: 500 });
+      }
+
+      // Forward the request to the service
+      const url = new URL(c.req.raw.url);
+      url.host = xGatewayServiceHost || service.host;
+
+      if (xGatewayServiceAuthType && !(xGatewayServiceAuthType in ServiceAuthType)) {
+        return c.json<Error>({
+          error: `x-gateway-service-auth-type header is invalid! It should be either of ${Object.values(ServiceAuthType).join(", ")}`
+        }, { status: 400 });
+      }
+
+      const token = xGatewayServiceToken || service.token;
+      const type = xGatewayServiceAuthType || service.authType;
+      const key = xGatewayServiceAuthKey || service.key;
+
+      if (type === ServiceAuthType.HEADER) {
+        const value = key.toLowerCase() === "authorization" ? `Bearer ${token}` : token;
+        c.req.raw.headers.append(key, value);
+      } else if (xGatewayServiceAuthType === ServiceAuthType.QUERY) {
+        url.searchParams.set(key, token);
+      }
+
+      const response = await fetch(url.toString(), c.req.raw);
```

Comment:
should we be reconstructing/copying the original request and stripping out the x-gateway headers?